api/utils/protobuf.go

```diff
@@ -0,0 +1,33 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/protoadapt"
+)
+
+// CloneProtoMsg returns a deep copy of msg. Modifying the returned
+// protobuf message will not affect msg.
+func CloneProtoMsg[T protoadapt.MessageV1](msg T) T {
+	// github.com/golang/protobuf/proto.Clone panics when trying to
+	// copy a map[K]V where the type of V is a slice of anything
+	// other than byte. See https://github.com/gogo/protobuf/issues/14
+	msgV2 := protoadapt.MessageV2Of(msg)
+	msgV2 = proto.Clone(msgV2)
+	return protoadapt.MessageV1Of(msgV2).(T)
```

Comment:
What sort of guarantee we have that this roundtrip will always yield the original message type? Should this be fallible?

Reply:
I just added a check to the test to validate that and it passes, not sure if that will always be the case for any protobuf message type though

Reply:
So far the unit tests are failing because some `nil` slice fields come back as an empty slice. That is technically different but not in a way that matters, do you agree?

Reply:
Or actually, the opposite of what I said

Reply:
I'm not sure that would help though, unless `CheckAndSetDefaults` of every message explicitly set empty slices to `nil`

Reply:
It should at least maintain parity with the original message even if it doesn't set the empty slices to nil. Most `CheckAndSetDefaults` do that IIRC.

Reply:
I don't think it's a big deal, just in terms of tests it would keep the parity of what is considered "empty" in the objects if that makes sense. I don't think it's absolutely required either. Feel free to disregard.

Reply:
Can we at least check that under the normal code paths for gogofast-generated structs we are guaranteed to get the same `T` out? Otherwise we risk trading one panic for another.

Reply:
What do you propose doing then? Adding a `bool` return and returning `false` if the type assertion fails, or adding tests for some of our generated messages with Copy methods?

Reply:
Even just a comment saying "after inspecting how the code works, this will not panic" would be fine - a unit test that tries to copy some of the specific messages that we want to copy would be great, too.

Reply:
Ok both suggestions are addressed